### ISSUE TITLE
Markdown popup (WIP)

### ISSUE THREAD

**Issue by dezza** [yegappan/lsp#66]
![vim9lsp-markdown](https://user-images.githubusercontent.com/402927/191416460-b4e78467-a886-49fc-8cad-aaef723a131e.png)

```vim
g:markdown_fenced_languages = ['help', 'python', 'bash=sh', 'vim', 'rust'] # add as needed
g:markdown_syntax_conceal = 1

set completeopt=menuone,popup,noinsert,popuphidden
def MarkdownPopup()
  w:vim9lsp_md_applied = get(w:, 'vim9lsp_md_applied', 0)
  var win = popup_findinfo()
  if w:vim9lsp_md_applied == 0
    var buf = winbufnr(win)
    if !!buf && buf != 1
      setwinvar(win, "&conceallevel", 3)
      setbufvar(buf, "&number", "0")
      setbufvar(buf, "&filetype", "markdown")
      w:vim9lsp_md_applied = 1
    endif
  endif
  popup_show(win)
enddef
autocmd CompleteChanged * MarkdownPopup()
```

I've been working on getting markdown syntax into the popup like the other lsp clients. I ended up using a similar strategy to how
[`complete-popuphidden`](https://vimhelp.org/insert.txt.html#complete-popuphidden) does it, to integrate with vim's own [`complete()`](https://vimhelp.org/builtin.txt.html#complete%28%29) function and `complete_info(dict)` https://github.com/yegappan/lsp/blob/7cc46b241b6e542fe0a8f2fafee65da3ddb1899b/autoload/lsp/handlers.vim#L292 (other clients uses `bufadd` and then `popup_create`).

There is still some things I'd like to improve. [vim-lsp](https://i.postimg.cc/3Nmd6Dy0/vim9lsp-markdown.png), for example unbinds the default markdown syntax regions (in the complete-popup that is) to achieve an even higher level of cleanliness. Mainly what it does is that it joins \`\`\` lang markdown-tokens (or code-fences) so they conceal on the same line (not leaving an empty line where \`\`\` was concealed). Also it defines its own fenced_languages, so your `g:markdown_fenced_languages` can stay agnostic and unassuming to your own `.md`-filetype preferences. I like this, I have already done it and saved lines that I can add to this function later, goes something like this
```vim
syntax sync clear
unlet g:markdown_fenced_languages
runtime! syntax/markdown.vim
let g:markdown_fenced_languages = ['help']
syntax clear markdownCodeBlock | syntax clear markdownCode
syntax region markdownCode matchgroup=Conceal start=/\%(``\)\@!`/ matchgroup=Conceal end=/\%(``\)\@!`/ containedin=TOP keepend concealends
syntax clear markdownEscape
syntax region markdownEscape matchgroup=markdownEscape start=/\\\ze[\\\x60*{}\[\]()#+\-,.!_>~|"$%&'\/:;<=?@^ ]/ end=/./ containedin=ALL keepend oneline concealends
```

Try it out and see if you encounter issues.

Would you prefer this as a PR for the `README.md` as customization or would you like me to attempt to integrate it with an option? Both are equally viable I guess, as there is no way to prematurely spice the complete-popup anyways without autocmd (that I know of).

PS.
I'm also going to attempt to make a function that searches `root_uri`-detecting files like `Cargo.toml`, `package.json` in parent-dirs when you open a file like `vim src/rust/test/test.rs` instead of `cd src/rust/test && vim test.rs` or `vim`, `:cd src/test` and `:e test.rs` (Notice the LSP server does not initialize properly without `cd` to workspace-dir first)

**Comment by dezza** [comment]
> Thanks for working on this. Can you create a PR with the changes to integrate this?

Right, my question was whether you think it would be better integrated to the main source or as commit to `README.md` ?

After its all done I don't think the necessity to integrate is larger.. The function will grow in size because of the mentioned markdown definitions.

An `autocmd User` hook PR with the current code suggested plus markdown tweaks is another possibility. It will make it possible to customize that function without forking.

**Comment by Shane-XB-Qian** [comment]
if you'd like to add this feature, i guess better be a 'au user',
since looks the doc format of lsp hover is not fixed, some lsp server return pure text, some just one markdown block, 
or specific to `rust` it is mixed:
![image](https://user-images.githubusercontent.com/13550529/192266480-f863f2b8-4f4e-490e-a647-a0f6a86a344d.png)
// so by 'au user' or by 'user local hacking' or whatever perhaps was better.

BTW: alternatively hover doc can be showed at `preview window` (someone had added this feature (option) into this lsp code).

**Comment by dezza** [comment]
@Shane-XB-Qian

I totally agree about customization and being unassuming is the best path :)

I will think "overriding" into the markdown-popup.

I have just gotten the hardest part; "Redefining markdown.vim"-syntax rules to work. So it conceals everything nicely, while retaining highlighting.

Also, unfortunately there's never a `kind` (plaintext|markdown) announced at initialize. It will have to be a flag that is toggled, say a hidden `lspserver._markdown_applied` property as soon as it gets the first markdown-reply it triggers, but only once! ☝🏻

Its all working, I only need to finish a single function and then test a few various servers to get a good overview of what to expect, cornercases or other potential issues.

**Comment by dezza** [comment]
Btw, currently I have made a `autoload/lsp/documentation.vim` import for the markdown features. I suppose this could house more documentation-specific features in the future.

How'd you like this idea? Or do you have a suggestion for better place to put it?

**Comment by dezza** [comment]
![m](https://user-images.githubusercontent.com/402927/193069796-479a86ce-6f87-4650-953b-4d67d40df97d.png)

Here's what it looks like now, pretty clean..

Here's how its made, I'll clean it up maybe merge some of those `win_execute` lines with a `|`
```vim
def MarkdownSyntaxCompact(win_id: number, buf_id: number)
  unsilent echomsg "MarkdownSyntaxCompact"
  win_execute(win_id, ':syntax sync clear')
  if exists('g:markdown_fenced_languages')
    var fenced_languages = g:markdown_fenced_languages
    unlet g:markdown_fenced_languages
    win_execute(win_id, ':runtime! syntax/markdown.vim')
    g:markdown_fenced_languages = fenced_languages
  else
    win_execute(win_id, ':runtime! syntax/markdown.vim')
  endif
  win_execute(win_id, ':unlet b:current_syntax')
  win_execute(win_id, ':syntax clear markdownCodeBlock')
  win_execute(win_id, ':syntax clear markdownCode')
  win_execute(win_id, ':syntax clear markdownEscape')
  win_execute(win_id, ':syntax region markdownCode matchgroup=Conceal start=/\%(``\)\@!`/ matchgroup=Conceal end=/\%(``\)\@!`/ containedin=TOP keepend concealends')
  win_execute(win_id, ':syntax region markdownIdDeclaration matchgroup=Conceal start=/\%(``\)\@!`/ end=/\%(``\)\@!`/ keepend concealends containedin=TOP')

  var marks = g:FindMarks(getbufline(buf_id, 0, '$')->join())
  for l in marks
    win_execute(win_id, $":syntax include @{toupper(l)} syntax/{l}.vim")
    win_execute(win_id, $"syntax region {toupper(l)} matchgroup=Conceal start=/```\\s*{l}\\s*/ matchgroup=Conceal end=/```$/ contains=@{toupper(l)} containedin=TOP keepend concealends")
  endfor
  win_execute(win_id, ':set conceallevel=3')
enddef
```

I'm gonna grep the source of as many lsp-servers I can find and see if I can find more improvements.

**Comment by rcasta74** [comment]
I've been working to get a similar result when I found this issue.
I'm trying to switch from coc.nvim to develop on java and I've found a couple of issues with completion:
1. not all possible item are displayed in the completion menu
2. documentation is missing for selected item

![image](https://user-images.githubusercontent.com/24571368/196057686-763becb0-6404-486f-ac0c-29fa126de8a4.png)

while using coc.nvim it look like

![image](https://user-images.githubusercontent.com/24571368/196057709-ee253b3b-1bc2-4dd0-8947-fe4ea9d85b71.png)

The first issue can be solved with `dup` property of `complete-item`, while for the missing documentation I need to get it in lazily way using [completionItem/resolve](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem_resolve) method.

With my [changes](https://github.com/rcasta74/lsp/commit/c874623b02676094687a3fd0567c980b1ba4b347) I was able to add completionItem/resolve method

![image](https://user-images.githubusercontent.com/24571368/196057949-ac847637-5139-48d2-9486-85f015421f28.png)

but the popup visualization is quite ugly:
* builtin markdown is not working correctly with the above text (it doesn't recognize the inner lists while it display them as code block)
* I wasn't able to force conceallevel to the popup
* the wrapping happen at the fixed width of the popup (and may happen in the middle of a word)

I will check if the above markdown code will help with java completion.

Could you give a look to my code if there may be any side effect on language server other than java while forcing to use lazy documentations? Or maybe I should add an option to use lazy doc or not.

@yegappan  If you feel that my code is good, I can open a PR only for completionItem/resolve method, without trying to force the markdown filetype to the popup, so for this issue we can then take care of also the cases where the documentation is not available early

**Comment by yegappan** [comment]
Can you open a PR for the completionItem/resolve method?  I will merge it.

**Comment by dezza** [comment]
@rcasta74 remember [conceallevel is local to window](https://vimhelp.org/options.txt.html#%27conceallevel%27), not global like other options. My last update 9 days ago here: https://github.com/dezza/lsp/commit/fc7dc7e33a4b975a44ea17f9cd27c5f9bba5c722

I don't know about the highlights/colors, but lots of these clients have their own tricks or way to present the docs.

**Comment by rcasta74** [comment]
> Can you open a PR for the completionItem/resolve method? I will merge it.

I have tried to add a check if the language server has the resolve capability ([here](https://github.com/rcasta74/lsp/commit/fce58d5d88aa7746d09bfaf0fe543928c622f87b) the new patch) and to apply the new code only in this case, but it doesn't work.
The issue is that in `AddFile` function, when it sets the options for insert mode completion, the lspserver is not yet ready, so the wrong options are applied (and the autocmd for `CompleteChanged` is not added). I have tried to add also `syncInit` property to the server but it didn't help.
The only working solution I have found so far is to add a loop waiting for the lspserver to be ready, but in this way the editor look like to be stuck for several seconds.
I will try to find an async solution.
Anyway there is another check that rely on the server to be ready (for the autoHighlight functionality), so the fix is required also for it.

Regarding the conceallevel, I missed to add '&' in front of the property name: now it works.

**Comment by Shane-XB-Qian** [comment]
1, #5 
cannot deploy it after vimenter, that's a knew problem.

2, #2 
if specific to java lsp (jdt from eclipse), there is `language/status`, maybe you can check the `ready` status by it.

3, pls check the `ProcessNotif` func in lsp/handlers.vim about that.

**Comment by yegappan** [comment]
>
> The issue is that in `AddFile` function, when it sets the options for
> insert mode completion, the lspserver is not yet ready, so the wrong
> options are applied (and the autocmd for `CompleteChanged` is not
> added). I have tried to add also `syncInit` property to the server but
> it didn't help.
>

You can add the autocmd when processing the LSP server initialize reply
message (in the ProcessInitializeReply function).  The document
highlight autocmd handler is added this way.

**Comment by yegappan** [comment]
I have tested your change with the python LSP server and it works.  Are you going to create a PR or can I merge the changes?

**Comment by dezza** [comment]
@yegappan https://github.com/yegappan/lsp/pull/75

**Comment by Shane-XB-Qian** [comment]
supposed Yegappan meant `completionItem/resolve` which truly required and useful specially java/jdt lsp, since for now no info doc when its compl.
// wish this can be merged. :-)

**Comment by rcasta74** [comment]
> You can add the autocmd when processing the LSP server initialize reply
> message (in the ProcessInitializeReply function). The document
> highlight autocmd handler is added this way.

I didn't noticed that all code relying to lspserver.caps were duplicated there.

I don't like duplicated code, so I tried to fix it (rcasta74@c9870ea). I think it should also resolve the TODO note I have removed.
I'm trying to improve it a bit more, but I'm not expert with vimscript so I'm taking long:
1) I have added a new autocmd in this way:
        `exe $'autocmd User LspServerReady{lspserver.name} ++once call BufferInit({bnr})'`
    I was trying to make it local to the buffer, but I cannot get `<buffer>` to work with User event.
2) I was trying to move the new function into buffer.vim, but I'm in trouble with some function belonging to a different script.

If you think the code is already good enough, I can open a PR as it is so that you can improve it yourself later.

After this I can open a PR for `completionItem/resolve` as well: I'm making some test together the above code and it seems to work quite well.

**Comment by yegappan** [comment]
Can you go ahead and open a PR?  I will merge the current version.  I can then make further changes on top of that and we can improve it.

**Comment by rcasta74** [comment]
Opened #76
As soon as it will be merged, I will open the other one

**Comment by yegappan** [comment]
The PR #76 is merged now.

**Comment by rcasta74** [comment]
completionItem/resolve #77

**Comment by yegappan** [comment]
The PR #77 is merged now.

**Comment by Shane-XB-Qian** [comment]
> completionItem/resolve #77
> Anyway the popup visualization is still to be improved, especially for the wrapping issue.

![image](https://user-images.githubusercontent.com/13550529/197407943-18352e5c-3fe2-4a46-95e7-ad593ec4b63d.png)

// a little flick and ugly... :-)  :rofl:

**Comment by dezza** [comment]
Is this `eclipse-jdt-ls` ? Last time I tried `java-language-server` I could not even resolve built-in packages created issue back then no fix.

**Comment by Shane-XB-Qian** [comment]
1, 'ugly' was due to unprintable chars specially conceal, removed them then it would be looks a bit fine.
2, 'flick' was due to async (to get the info doc), this looks had no good solution since `async` is required here.

**Comment by Shane-XB-Qian** [comment]
![image](https://user-images.githubusercontent.com/13550529/197456849-091e3bdd-5587-4007-8544-89ee5e5cd7fc.png)

**Comment by yegappan** [comment]
With the recent changes to support markdown syntax in the documentation completion popup window, can this be closed?
